lib/js-yaml/dumper.js

```diff
@@ -150,81 +174,331 @@ function testImplicitResolving(state, str) {
+  if (this.source.length > this.checkpoint) {
+    this.takeUpTo(this.source.length);
+  }
+};
+
+
+function writeScalar (state, object, level) {
+  if (0 === object.length) {
+    state.dump = "''";
+    return;
+  }
 
   if (-1 !== DEPRECATED_BOOLEANS_SYNTAX.indexOf(object)) {
-    // Ensure compatibility with YAML 1.0/1.1 loaders.
-    isQuoted = true;
-  } else if (0 === object.length) {
-    // Quote empty string
-    isQuoted = true;
-  } else if (CHAR_SPACE    === first ||
-             CHAR_SPACE    === object.charCodeAt(object.length - 1)) {
-    isQuoted = true;
-  } else if (CHAR_MINUS    === first ||
-             CHAR_QUESTION === first) {
-    // Don't check second symbol for simplicity
-    isQuoted = true;
-  }
-
-  for (position = 0, length = object.length; position < length; position += 1) {
-    character = object.charCodeAt(position);
-
-    if (!isQuoted) {
-      if (CHAR_TAB                  === character ||
-          CHAR_LINE_FEED            === character ||
-          CHAR_CARRIAGE_RETURN      === character ||
-          CHAR_COMMA                === character ||
-          CHAR_LEFT_SQUARE_BRACKET  === character ||
-          CHAR_RIGHT_SQUARE_BRACKET === character ||
-          CHAR_LEFT_CURLY_BRACKET   === character ||
-          CHAR_RIGHT_CURLY_BRACKET  === character ||
-          CHAR_SHARP                === character ||
-          CHAR_AMPERSAND            === character ||
-          CHAR_ASTERISK             === character ||
-          CHAR_EXCLAMATION          === character ||
-          CHAR_VERTICAL_LINE        === character ||
-          CHAR_GREATER_THAN         === character ||
-          CHAR_SINGLE_QUOTE         === character ||
-          CHAR_DOUBLE_QUOTE         === character ||
-          CHAR_PERCENT              === character ||
-          CHAR_COMMERCIAL_AT        === character ||
-          CHAR_COLON                === character ||
-          CHAR_GRAVE_ACCENT         === character) {
-        isQuoted = true;
+    state.dump = "'" + object + "'";
+    return;
+  }
+
+  var simple = true;
+  var first = object.length ? object.charCodeAt(0) : 0;
+  var spaceWrap = (CHAR_SPACE === first ||
+                   CHAR_SPACE === object.charCodeAt(object.length - 1));
+
+  var maybeNum = (CHAR_MINUS    === first ||
+                  CHAR_QUESTION === first);
+
+  // can only use > and | if not wrapped in spaces.
+  var folded, literal;
+  if (spaceWrap) {
+    simple = false;
+    folded = false;
+    literal = false;
+  } else {
+    folded = true;
+    literal = true;
+  }
+
+  if (maybeNum) {
+    simple = false;
+  }
+
+  var single = true;
+  var double = new StringBuilder(object);
+
+  var sawLineFeed = false;
+  var linePosition = 0;
+  var longestLine = 0;
+
+  var indent = state.indent * level;
+  var max = 80;
+  if (indent < 40) {
+    max -= indent;
+  } else {
+    max = 40;
+  }
+
+  for (var position = 0; position < object.length; position++) {
```

Comment:
All `var` statements must be declared at top-level of function body. Also it's preferred, but not required to declare them at very beginning of the function body.